### PR TITLE
[Execution] Improve, add tests, and docs to hyper params and run context

### PR DESCRIPTION
**Pull request by yaronha**
* support child (iteration) context and custom hyper-params 
* make run context a python context (support with)
* add docs and examples to run context
* add tests to all hyper param options
* fixes ans better exception handling

### CODE REVIEW

File: mlrun/datastore/inmem.py [hunk 1]
```diff
@@ -52,5 +56,27 @@ def stat(self, key):
     def listdir(self, key):
         return []
 
-    def as_df(self, key, columns=None, df_module=None, format="", **kwargs):
-        return self._get_item(key)
+    def as_df(self, url, subpath, columns=None, df_module=None, format="", **kwargs):
+        item = self._get_item(subpath)
+        if hasattr(item, "read_csv"):
+            return item
```
Comment (Hedingber) [inline mlrun/datastore/inmem.py:62]: what's the logic behind this ?

File: mlrun/execution.py [hunk 2]
```diff
@@ -96,61 +98,107 @@ def __init__(self, autocommit=False, tmp="", log_stream=None):
...
+    def update_child_iterations(self, best_run=0, commit_children=False):
+        """update children results in the parent, and optionally mark the best
+
+        :param best_run:  marks the child iteration number (starts from 1)
+        :param commit_children:  commit all child runs to the db
+        """
+        if not self._children:
+            return
+        if commit_children:
+            for c in self._children:
```
Comment (Hedingber) [inline mlrun/execution.py:167]:
```suggestion
            for child in self._children:
```

File: tests/test_run.py [hunk 3]
```diff
@@ -142,6 +148,72 @@ def test_handler_hyperlist():
...
+    grid_params = '{"p2": [2,1,3], "p3": [10,20]}'
+    mlrun.datastore.set_in_memory_item("params.json", grid_params)
+
+    run_spec = tag_test(base_spec, "test_hyper_grid")
+    run_spec.with_param_file("memory://params.json", selector="r1", strategy="grid")
+    run = new_function().run(run_spec, handler=hyper_func)
+
+    verify_state(run)
+    # 3 x p2, 2 x p3 = 6 iterations
+    assert len(run.status.iterations) == 1 + 2 * 3, "wrong number of iterations"
```
Comment (Hedingber) [inline tests/test_run.py:161]: 1 + 2*3 is 7 not 6

File: mlrun/execution.py [hunk 4]
```diff
@@ -587,9 +713,7 @@ def log_model(
         """
 
         if training_set is not None and inputs:
-            raise mlrun.errors.MLRunInvalidArgumentError(
-                "cannot specify inputs and training set together"
-            )
+            raise MLCtxValueError("cannot specify inputs and training set together")
```
Comment (Hedingber) [inline mlrun/execution.py:716]: What's the idea behind this `MLCtxValueError` ? 
It doesn't even inherit from `ValueError`... I don't see a reason not using `MLRunInvalidArgumentError` everywhere

**Review by Hedingber — CHANGES_REQUESTED** [review]

File: mlrun/datastore/inmem.py (hunk 1 above)
Comment (yaronha) [reply mlrun/datastore/inmem.py:62]: Detect if it's a dataframe/dask/etc..

File: mlrun/execution.py (hunk 2 above)
Comment (yaronha) [reply mlrun/execution.py:167]: Ok

File: tests/test_run.py (hunk 3 above)
Comment (yaronha) [reply tests/test_run.py:161]: The +1 is the header line

File: mlrun/execution.py (hunk 4 above)
Comment (yaronha) [reply mlrun/execution.py:716]: This is what we had before, I'm not a fan of sticking MLRunInvalidArgumentError everywhere, why don't we just use the built in ValueError?

**Review by Hedingber — APPROVED** [review]

**Pull request merged**